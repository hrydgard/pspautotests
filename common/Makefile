TARGET_LIB = libcommon.a
OBJS = common.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak
