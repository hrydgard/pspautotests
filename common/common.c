#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <assert.h>
#include <pspdebug.h>
#include <pspthreadman.h>
#include <psploadexec.h>
#include <pspctrl.h>
#include <pspdisplay.h>
#include <pspiofilemgr.h>
//#include <pspkdebug.h>

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <sys/lock.h>
#include <sys/fcntl.h>

//#include "local.h"

#include "sysmem-imports.h"

/*
enum PspThreadAttributes
{
	PSP_THREAD_ATTR_VFPU = 0x00004000,
	PSP_THREAD_ATTR_USER = 0x80000000,
	PSP_THREAD_ATTR_USBWLAN = 0xa0000000,
	PSP_THREAD_ATTR_VSH = 0xc0000000,
	PSP_THREAD_ATTR_SCRATCH_SRAM = 0x00008000,
	PSP_THREAD_ATTR_NO_FILLSTACK = 0x00100000,
	PSP_THREAD_ATTR_CLEAR_STACK = 0x00200000,
};
*/
/*
enum PspModuleInfoAttr
{
	PSP_MODULE_USER			= 0,
	PSP_MODULE_NO_STOP		= 0x0001,
	PSP_MODULE_SINGLE_LOAD	= 0x0002,
	PSP_MODULE_SINGLE_START	= 0x0004,
	PSP_MODULE_KERNEL		= 0x1000,
};
*/

PSP_MODULE_INFO("TESTMODULE", PSP_MODULE_USER, 1, 0);
PSP_MAIN_THREAD_ATTR(PSP_THREAD_ATTR_USER | PSP_THREAD_ATTR_VFPU);
//PSP_MAIN_THREAD_ATTR(PSP_THREAD_ATTR_USER);

#define EMULATOR_DEVCTL__GET_HAS_DISPLAY 0x00000001
#define EMULATOR_DEVCTL__SEND_OUTPUT     0x00000002
#define EMULATOR_DEVCTL__IS_EMULATOR     0x00000003
#define EMULATOR_DEVCTL__SEND_CTRLDATA   0x00000010
#define EMULATOR_DEVCTL__EMIT_SCREENSHOT 0x00000020

unsigned int RUNNING_ON_EMULATOR = 0;
unsigned int CHECKPOINT_ENABLE_TIME = 0;
unsigned int CHECKPOINT_OUTPUT_DIRECT = 0;
unsigned int HAS_DISPLAY = 1;

// 21 MB to give space for thread stacks and etc.
unsigned int sce_newlib_heap_kb_size = 21504;

extern int test_main(int argc, char *argv[]);

FILE stdout_back = {NULL};
//int KprintfFd = 0;

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	if (CHECKPOINT_OUTPUT_DIRECT) {
		// This is easier to debug in the emulator, but printf() reschedules on the real PSP.
		vprintf(format, args);
	} else {
		schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	}
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

SceUID reschedThread;
volatile int didResched = 0;
int reschedFunc(SceSize argc, void *argp) {
	didResched = 1;
	return 0;
}

u64 lastCheckpoint = 0;
void checkpoint(const char *format, ...) {
	u64 currentCheckpoint = sceKernelGetSystemTimeWide();
	if (CHECKPOINT_ENABLE_TIME) {
		schedf("[%s/%lld] ", didResched ? "r" : "x", currentCheckpoint - lastCheckpoint);
	} else {
		schedf("[%s] ", didResched ? "r" : "x");
	}

	sceKernelTerminateThread(reschedThread);

	if (format != NULL) {
		va_list args;
		va_start(args, format);
		if (CHECKPOINT_OUTPUT_DIRECT) {
			// This is easier to debug in the emulator, but printf() reschedules on the real PSP.
			vprintf(format, args);
		} else {
			schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
		}
		va_end(args);
	}

	didResched = 0;
	sceKernelStartThread(reschedThread, 0, NULL);

	if (format != NULL) {
		schedf("\n");
	}
	
	lastCheckpoint = currentCheckpoint;
}

void checkpointNext(const char *title) {
	if (schedfBufferPos != 0) {
		schedf("\n");
	}
	flushschedf();
	didResched = 0;
	if (title != NULL)
		checkpoint(title);
}

static int writeStdoutHook(struct _reent *ptr, void *cookie, const char *buf, int buf_len) {
	char temp[1024 + 1];

	//if (KprintfFd > 0) sceIoWrite(KprintfFd, buf, buf_len);
	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("emulator:", EMULATOR_DEVCTL__SEND_OUTPUT, (void *)buf, buf_len, NULL, 0);
	}

	if (buf_len < sizeof(temp)) {
		if (HAS_DISPLAY) {
			memcpy(temp, buf, buf_len);
			temp[buf_len] = 0;

			//Kprintf("%s", temp);
			pspDebugScreenPrintf("%s", temp);
		}
	}
	
	if (stdout_back._write != NULL) {
		return stdout_back._write(ptr, cookie, buf, buf_len);
	} else {
		return buf_len;
	}
}

typedef int (*SdkVerFunc)(u32 ver);
typedef struct SdkVerFuncTable {
	u32 id;
	SdkVerFunc func;
} SdkVerFuncTable;

static SdkVerFuncTable sdkVerFuncs[] = {
	{0, sceKernelSetCompiledSdkVersion},
	{370, sceKernelSetCompiledSdkVersion370},
	{380, sceKernelSetCompiledSdkVersion380_390},
	{395, sceKernelSetCompiledSdkVersion395},
	{401, sceKernelSetCompiledSdkVersion401_402},
	{500, sceKernelSetCompiledSdkVersion500_505},
	{507, sceKernelSetCompiledSdkVersion507},
	{600, sceKernelSetCompiledSdkVersion600_602},
	{603, sceKernelSetCompiledSdkVersion603_605},
	{606, sceKernelSetCompiledSdkVersion606},
};

static void updateSdkVer(int argc, char *argv[]) {
	int i = 0;
	u32 ver = 0xFFFFFFFF;
	u32 funcID = 0;
	for (i = 1; i < argc; ++i) {
		if (!strncmp(argv[i], "--sdkver=", strlen("--sdkver="))) {
			ver = strtoul(argv[i] + strlen("--sdkver="), NULL, 16);
		}
		if (!strncmp(argv[i], "--sdkver-func=", strlen("--sdkver-func="))) {
			funcID = strtol(argv[i] + strlen("--sdkver-func="), NULL, 10);
		}
	}

	SdkVerFunc func = NULL;
	for (i = 0; i < sizeof(sdkVerFuncs) / sizeof(sdkVerFuncs[0]); ++i) {
		if (sdkVerFuncs[i].id == funcID) {
			func = sdkVerFuncs[i].func;
		}
	}

	if (func == NULL) {
		fprintf(stderr, "Unknown sdkver-func value.\n");
		exit(1);
	}

	if (ver != 0xFFFFFFFF) {
		if (func(ver) != 0) {
			printf("WARNING: Setting sdkver returned failure.\n");
		}
	}
}

void test_begin() {
	if (HAS_DISPLAY) {
		pspDebugScreenInit();
	}

	if (RUNNING_ON_EMULATOR && !HAS_DISPLAY) {
		fclose(stdout);
		stdout = fmemopen(alloca(4), 4, "wb");
		stdout_back._write = NULL;
		
		//stderr = stdout;
		setbuf(stdout, NULL);
	} else {
    // Send the output to the host.
		freopen("host0:/__testoutput.txt", "wb", stdout);
		freopen("host0:/__testerror.txt", "wb", stderr);
		stdout_back._write = stdout->_write;
	}
	stdout->_write = writeStdoutHook;

	setvbuf(stdout, NULL, _IONBF, 0);
	setvbuf(stderr, NULL, _IONBF, 0);
	
	setbuf(stderr, NULL);

	reschedThread = sceKernelCreateThread("resched", &reschedFunc, sceKernelGetThreadCurrentPriority(), 0x1000, 0, NULL);
}

void test_end() {
	flushschedf();

	fflush(stdout);
	fflush(stderr);
	
	fclose(stdout);
	fclose(stderr);

	if (!RUNNING_ON_EMULATOR) {
		FILE *finish = fopen("host0:/__testfinish.txt", "wb");
		if (finish)
		{
			fwrite("1", sizeof(char), 1, finish);
			fclose(finish);
		}
	}

  // Disabled the wait, much more convienent when running automated.
	if (0 && !RUNNING_ON_EMULATOR) {
		SceCtrlData key;
		while (1) {
			sceCtrlReadBufferPositive(&key, 1);
			if (key.Buttons & PSP_CTRL_CROSS) break;
		}
	}
	
	//fclose(stdout);
	sceKernelExitGame();
	
	exit(0);
}

int test_psp_exit_callback(int arg1, int arg2, void *common) {
	exit(0);
	return 0;
}

int test_psp_callback_thread(SceSize args, void *argp) {
	int cbid;
	cbid = sceKernelCreateCallback("Exit Callback", test_psp_exit_callback, NULL);
	sceKernelRegisterExitCallback(cbid);
	sceKernelSleepThreadCB();
	return 0;
}

int test_psp_setup_callbacks(void) {
	int thid = 0;
	thid = sceKernelCreateThread("update_thread",  test_psp_callback_thread, 0x11, 0xFA0, 0, 0);
	if (thid >= 0) sceKernelStartThread(thid, 0, 0);
	return thid;
}

//#define START_WITH "ms0:/PSP/GAME/virtual"

/*
void emitInt(int v) {
	asm("syscall 0x1010");
}

void emitFloat(float v) {
	asm("syscall 0x1011");
}

void emitString(char *v) {
	asm("syscall 0x1012");
}

void emitComment(char *v) {
	asm("syscall 0x1012");
}

void emitMemoryBlock(void *address, unsigned int size) {
	asm("syscall 0x1013");
}

void emitHex(void *address, unsigned int size) {
	asm("syscall 0x1014");
}
*/

unsigned char bmpHeader[54] = {
	0x42, 0x4D, 0x38, 0x80, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x36, 0x00,
	0x00, 0x00, 0x28, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x10, 0x01,
	0x00, 0x00, 0x01, 0x00, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x80,
	0x08, 0x00, 0x12, 0x0B, 0x00, 0x00, 0x12, 0x0B, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};


uint extractBits(uint value, int offset, int size) {
	return (value >> offset) & ((1 << size) - 1);
}

void emulatorEmitScreenshot() {
	int file;

	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("kemulator:", EMULATOR_DEVCTL__EMIT_SCREENSHOT, NULL, 0, NULL, 0);
	}
	else
	{
		uint topaddr;
		int bufferwidth;
		int pixelformat;

		sceDisplayGetFrameBuf((void **)&topaddr, &bufferwidth, &pixelformat, 0);
		
        if (topaddr & 0x80000000) {
            topaddr |= 0xA0000000;
        } else {
            topaddr |= 0x40000000;
        }
	
		if ((file = sceIoOpen("host0:/__screenshot.bmp", PSP_O_CREAT | PSP_O_WRONLY | PSP_O_TRUNC, 0777)) >= 0) {
			int y, x;
			uint c;
			uint* vram_row;
			uint* row_buf = (uint *)malloc(512 * 4);
			sceIoWrite(file, &bmpHeader, sizeof(bmpHeader));
			for (y = 0; y < 272; y++) {
				vram_row = (uint *)(topaddr + 512 * 4 * (271 - y));
				for (x = 0; x < 512; x++) {
					c = vram_row[x];
					/*
					row_buf[x] = (
						((extractBits(c,  0, 8)) <<  0) |
						((extractBits(c,  8, 8)) <<  8) |
						((extractBits(c, 16, 8)) << 16) |
						((                0x00 ) << 24) |
					0);
					*/
					row_buf[x] = (
						((extractBits(c, 16, 8)) <<  0) |
						((extractBits(c,  8, 8)) <<  8) |
						((extractBits(c,  0, 8)) << 16) |
						((                0x00 ) << 24) |
					0);
				}
				sceIoWrite(file, row_buf, 512 * 4);
			}
			free(row_buf);
			//sceIoWrite(file, (void *)topaddr, bufferwidth * 272 * 4);
			//sceIoFlush();
			sceIoClose(file);
		}
	}
}

void emulatorSendSceCtrlData(SceCtrlData* pad_data) {
	sceIoDevctl("kemulator:", EMULATOR_DEVCTL__SEND_CTRLDATA, pad_data, sizeof(SceCtrlData), NULL, 0);
}

int main(int argc, char *argv[]) {
	int retval = 0;

	//KprintfFd = sceIoOpen("emulator:/Kprintf", O_WRONLY, 0777);
	//RUNNING_ON_EMULATOR = (KprintfFd > 0);
	
	if (sceIoDevctl("kemulator:", EMULATOR_DEVCTL__IS_EMULATOR, NULL, 0, NULL, 0) == 0) {
		RUNNING_ON_EMULATOR = 1;
	}
	
	// TEMP HACK
	//RUNNING_ON_EMULATOR = 0;

	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("kemulator:", EMULATOR_DEVCTL__GET_HAS_DISPLAY, NULL, 0, &HAS_DISPLAY, sizeof(HAS_DISPLAY));
	}
	
	#ifdef GRAPHIC_TEST
		HAS_DISPLAY = 0;
	#endif

	//if (strncmp(argv[0], START_WITH, strlen(START_WITH)) == 0) RUNNING_ON_EMULATOR = 1;

	if (!RUNNING_ON_EMULATOR) {
		test_psp_setup_callbacks();
	}
	atexit(sceKernelExitGame);

	test_begin();
	{
		pspDebugScreenPrintf("RUNNING_ON_EMULATOR: %s - %s\n", RUNNING_ON_EMULATOR ? "yes" : "no", argv[0]);
		updateSdkVer(argc, argv);

		retval = test_main(argc, argv);
	}
	test_end();
	
	return retval;
}
