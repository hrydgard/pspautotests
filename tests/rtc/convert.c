#include <common.h>
#include "rtc_common.h"


void checkRtcConvertLocalTimeToUTC()
{
	printf("Checking sceRtcConvertLocalTimeToUTC\n");
	u64 tick1=62135596800000000ULL;
	u64 tick2=0; 
	sceRtcConvertLocalTimeToUTC(&tick1, &tick2);
	printf("epoch : %llu\n", tick1);
	printf("epoch as UTC: %llu\n", tick2);
}

void checkRtcConvertUtcToLocalTime()
{
	printf("Checking checkRtcConvertUtcToLocalTime\n");
	u64 tick1=62135596800000000ULL;
	u64 tick2=0; 
	sceRtcConvertUtcToLocalTime(&tick1, &tick2);
	printf("epoch as UTC: %llu\n", tick1);
	printf("epoch as UTC as Local : %llu\n", tick2);
}

void checkRtcSetTime_t()
{
	printf("Checking sceRtcSetTime_t\n");

	pspTime pt;
	printf("from 0:%d\n", sceRtcSetTime_t(&pt, 0));
	DumpPSPTime("", &pt);
	printf("from epoc:%d\n",sceRtcSetTime_t(&pt, 62135596800ULL));
	DumpPSPTime("", &pt);
	printf("from epoc(64):%d\n",sceRtcSetTime64_t(&pt, 62135596800000000ULL));
	DumpPSPTime("", &pt);
	printf("from 2012, 9, 20, 7, 12, 15, 500:%d\n",sceRtcSetTime_t(&pt, 1348125135));
	DumpPSPTime("", &pt);

	printf("from epoc&0xffffffff:%d\n",sceRtcSetTime_t(&pt, 62135596800ULL&0xffffffff));
	DumpPSPTime("", &pt);
}

void checkRtcGetTime_t()
{
	printf("Checking sceRtcGetTime_t\n");
	pspTime pt;
	u64 ticks=0;
	FillPSPTime(&pt,2012,9,20,7,12,15,500);
	printf("from epoc:%d\n",sceRtcGetTime_t( &pt, &ticks));
	DumpPSPTime("", &pt);
	printf("ticks: %llu\n", ticks);
	printf("from epoc (64):%d\n",sceRtcGetTime64_t( &pt, &ticks));
	DumpPSPTime("", &pt);
	printf("ticks: %llu\n", ticks);
	sceRtcSetTime_t(&pt, ticks);
	DumpPSPTime("time_t time : ",&pt);
}

void checkRtcSetDosTime()
{
	printf("Checking sceRtcSetDosTime\n");

	pspTime pt;
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 0));
	DumpPSPTime("0 = ",&pt);
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 1));
	DumpPSPTime("1 = ",&pt);
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 10));
	DumpPSPTime("10 = ",&pt);
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 100));
	DumpPSPTime("100 = ",&pt);
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 1000));
	DumpPSPTime("1000 = ",&pt);
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 10000));
	DumpPSPTime("10000 = ",&pt);
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 100000));
	DumpPSPTime("100000 = ",&pt);
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 1000000));
	DumpPSPTime("1000000 = ",&pt);
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 10000000));
	DumpPSPTime("10000000 = ",&pt);
	printf("from epoc:%d\n",sceRtcSetDosTime(&pt, 62135596800000000ULL));
	DumpPSPTime("62135596800000000ULL = ",&pt);
}


// Getting warnings but it seems sceRtcGetDosTime is simply declared wrong
void checkRtcGetDosTime()
{
	printf("Checking sceRtcGetDosTime\n");

	pspTime pt;
	u64 ticks=0;
	FillPSPTime(&pt, 2107, 9, 11, 24, 0, 0, 0);
	printf("from epoc:%d\n",sceRtcGetDosTime( &pt, &ticks));
	DumpPSPTime("", &pt);
	printf("ticks: %llu\n", ticks);

	// false date
	pt.year = 1979;
	printf("from epoc:%d\n",sceRtcGetDosTime( &pt, &ticks));
	pt.year = 2108;
	printf("from epoc:%d\n",sceRtcGetDosTime( &pt, &ticks));

}

void checkRtcSetWin32FileTime()
{
	pspTime pt;
	u64 ft;
	int result;
	memset(&pt, 0, sizeof(pt));

	printf("Checking sceRtcSetWin32FileTime\n");
	// Crash.
	//printf("  NULL args: %08x\n", sceRtcSetWin32FileTime(NULL, NULL));
	ft = 0;
	result = sceRtcSetWin32FileTime(&pt, ft);
	printf("  Zero time (1601 January 01): (%08x)", result);
	DumpPSPTime("", &pt);
	ft = 127779156600000010ULL;
	result = sceRtcSetWin32FileTime(&pt, ft);
	printf("  Arbitrary date/time: (%08x)", result);
	DumpPSPTime("", &pt);

	// This gives weird results.  Games unlikely to call it, so let's ignore for now.
	//ft = -(365ULL * 86400ULL * 10000000ULL);
	//result = sceRtcSetWin32FileTime(&pt, ft);
	//printf("  Before 1601 January 01: (%08x)", result);
	//DumpPSPTime("", &pt);
	printf("\n");
}

void checkRtcGetWin32FileTime()
{
	pspTime pt;
	u64 ft = -1337;
	int result;

	printf("Checking sceRtcGetWin32FileTime\n");

	// It's not clear, but it seems like sceRtcGetWin32FileTime() fails to return error messages
	// properly if it hasn't processed at least one error properly.
	// Trying to emulate this isn't necessarily a bad idea, but it's hard to determine its rules...
	FillPSPTime(&pt, 1600, 1, 1, 0, 0, 0, 0);
	sceRtcGetWin32FileTime(&pt, &ft);

	// Crash.
	//printf("  NULL args: %08x\n", sceRtcGetWin32FileTime(NULL, NULL));
	ft = -1337;
	result = sceRtcGetWin32FileTime(&pt, NULL);
	printf("  NULL filetime: %lld (%08x)\n", ft, result);
	memset(&pt, 0, sizeof(pt));
	ft = -1337;
	result = sceRtcGetWin32FileTime(&pt, &ft);
	printf("  Zeroed time: %lld (%08x)\n", ft, result);
	FillPSPTime(&pt, 2005, 11, 31, 13, 01, 00, 1);
	ft = -1337;
	result = sceRtcGetWin32FileTime(&pt, &ft);
	printf("  Arbitrary date/time: %lld (%08x)\n", ft, result);
	FillPSPTime(&pt, 1601, 1, 1, 0, 0, 0, 0);
	ft = -1337;
	result = sceRtcGetWin32FileTime(&pt, &ft);
	printf("  1601 January 01: %lld (%08x)\n", ft, result);
	FillPSPTime(&pt, 1600, 1, 1, 0, 0, 0, 0);
	ft = -1337;
	result = sceRtcGetWin32FileTime(&pt, &ft);
	printf("  1600 January 01: %lld (%08x)\n", ft, result);
	FillPSPTime(&pt, 0, 1, 1, 0, 0, 0, 0);
	ft = -1337;
	result = sceRtcGetWin32FileTime(&pt, &ft);
	printf("  1 January 01: %lld (%08x)\n", ft, result);
	printf("\n");
}

void checkSetTick()
{
	pspTime pt;
	u64 ticks = 835072;

	memset(&pt, 0, sizeof(pt));

	printf("checkSetTick: empty small value: %08x\n", sceRtcSetTick(&pt, &ticks));
	DumpPSPTime("",&pt);

	ticks = 62135596800000000ULL;
	memset(&pt, 0, sizeof(pt));
	printf("checkSetTick: empty rtcMagicOffset: %08x\n", sceRtcSetTick(&pt, &ticks));
	DumpPSPTime("",&pt);

	FillPSPTime(&pt,2012,9,20,7,12,15,500);
	printf("Normal: %08x\n", sceRtcGetTick(&pt, &ticks)); // if this does depend on timezone the next bit might have differnt results
	printf("checkSetTick: 2012, 09, 20, 7, 12, 15, 500: %08x\n", sceRtcSetTick(&pt, &ticks));
	DumpPSPTime("",&pt);


	FillPSPTime(&pt,2010,9,20,7,12,15,500);
	printf("preset\n");
	DumpPSPTime("",&pt);
	printf("checkSetTick: not empty:%08x\n", sceRtcSetTick(&pt, &ticks));
	DumpPSPTime("",&pt);
}

void checkGetTick() {
	pspTime pt;
	u64 ticks;

	printf("Checking sceRtcGetTick\n");

	FillPSPTime(&pt,10,1,1,0,0,0,0);
	printf("Normal: %08x\n", sceRtcGetTick(&pt, &ticks));
	printf("Ticks : %llu\n",ticks);

	FillPSPTime(&pt,9998,1,1,0,0,0,0);
	printf("Bad date: %08x\n", sceRtcGetTick(&pt, &ticks));
	printf("Ticks : %llu\n",ticks);

	FillPSPTime(&pt,0,1,1,0,0,0,0);
	printf("Min year: %08x\n", sceRtcGetTick(&pt, &ticks));
	printf("Ticks : %llu\n",ticks);

	FillPSPTime(&pt,9999,1,1,0,0,0,0);
	printf("Max year: %08x\n", sceRtcGetTick(&pt, &ticks));
	printf("Ticks : %llu\n",ticks);

	FillPSPTime(&pt,10000,1,1,0,0,0,0);
	printf("Year overflow: %08x\n", sceRtcGetTick(&pt, &ticks));
	printf("Ticks : %llu\n",ticks);
}

void checkRFC2822() {
	pspTime pt;
	u64 ticks;
	char buffer[256];

	FillPSPTime(&pt,2012,1,30,2,12,15,900);
	sceRtcGetTick(&pt, &ticks);
	printf("Normal:\n");
	DumpPSPTime("",&pt);
	sceRtcFormatRFC2822(buffer, &ticks, 0);
	printf("RFC 2822: %s\n", buffer);

	FillPSPTime(&pt,2010,9,20,7,12,15,500);
	sceRtcGetTick(&pt, &ticks);
	printf("Normal:\n");
	DumpPSPTime("",&pt);
	sceRtcFormatRFC2822(buffer, &ticks, 0);
	printf("RFC 2822: %s\n", buffer);
}

void checkRFC3339() {
	pspTime pt;
	u64 ticks;
	char buffer[256];

	FillPSPTime(&pt,2012,1,30,2,12,15,900);
	sceRtcGetTick(&pt, &ticks);
	printf("Normal:\n");
	DumpPSPTime("",&pt);
	sceRtcFormatRFC3339(buffer, &ticks, 0);
	printf("RFC 3339: %s\n", buffer);

	FillPSPTime(&pt,2010,9,20,7,12,15,500);
	sceRtcGetTick(&pt, &ticks);
	printf("Normal:\n");
	DumpPSPTime("",&pt);
	sceRtcFormatRFC3339(buffer, &ticks, 0);
	printf("RFC 3339: %s\n", buffer);
}

void testParse(const char *title, const char *dateString) {
	u64 ticks = 0;
	int result = sceRtcParseDateTime(&ticks, dateString);

	char buffer[256] = {0};
	sceRtcFormatRFC3339(buffer, &ticks, 0);
	printf("  %s: %08x: %s -> %s\n", title, result, dateString, buffer);
}

void checkRtcParseDateTime() {
	u64 ticks;
	char buffer[256];

	printf("\nsceRtcParseDateTime:\n");

	testParse("From RFC3339 (Z)", "2010-09-20T07:12:15.00Z");
	testParse("From RFC3339 (offset)", "2010-09-20T07:12:15.00+01:00");
	testParse("From RFC3339 (no ms)", "2010-09-20T07:12:15Z");
	testParse("Bad RFC3339 (missing TZ)", "2010-09-20T07:12:15");
	testParse("Space separator", "2010-09-20 07:12:15Z");
	testParse("Human speak", "today");
	testParse("Time only", "07:12:15");
	testParse("Date only", "2010-09-20");
	testParse("From RFC2822", "Mon, 30 Jan 2012 02:12:15 +0000");
	testParse("From RFC2822 date", "Mon, 30 Jan 2012");
	testParse("From RFC2822 full weekday", "Monday, 30 Jan 2012 02:12:15 +0000");
	testParse("From RFC2822 no weekday", "30 Jan 2012 02:12:15 +0000");
	testParse("From RFC2822 + spaces", "Mon,  30  Jan  2012  02:12:15  +0000");
	testParse("Blank", "");
	// Crashes.
	//testParse("NULL", NULL);
}

int main(int argc, char **argv) {
	checkRtcConvertLocalTimeToUTC();
	checkRtcConvertUtcToLocalTime();

	checkRtcSetTime_t();
	checkRtcGetTime_t();
	checkRtcSetDosTime();
	checkRtcGetDosTime();
	checkRtcSetWin32FileTime();
	checkRtcGetWin32FileTime();

	checkGetTick();
	checkSetTick();	

	checkRFC2822();
	checkRFC3339();
	
	checkRtcParseDateTime();

	return 0;
}