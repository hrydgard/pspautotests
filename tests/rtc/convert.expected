Checking sceRtcConvertLocalTimeToUTC
epoch : 62135596800000000
epoch as UTC: 62135593200000000
Checking checkRtcConvertUtcToLocalTime
epoch as UTC: 62135596800000000
epoch as UTC as Local : 62135600400000000
Checking sceRtcSetTime_t
from 0:0
 1970, 1, 1, 0, 0, 0, 0
from epoc:0
 2033, 7, 27, 5, 24, 16, 0
from epoc(64):0
 54861, 7, 13, 12, 29, 40, 441792
from 2012, 9, 20, 7, 12, 15, 500:0
 2012, 9, 20, 7, 12, 15, 0
from epoc&0xffffffff:0
 2033, 7, 27, 5, 24, 16, 0
Checking sceRtcGetTime_t
from epoc:0
 2012, 9, 20, 7, 12, 15, 500
ticks: 1348125135
from epoc (64):0
 2012, 9, 20, 7, 12, 15, 500
ticks: 1348125135
time_t time :  2012, 9, 20, 7, 12, 15, 0
Checking sceRtcSetDosTime
from epoc:0
0 =  1980, 0, 0, 0, 0, 0, 0
from epoc:0
1 =  1980, 0, 0, 0, 0, 2, 0
from epoc:0
10 =  1980, 0, 0, 0, 0, 20, 0
from epoc:0
100 =  1980, 0, 0, 0, 3, 8, 0
from epoc:0
1000 =  1980, 0, 0, 0, 31, 16, 0
from epoc:0
10000 =  1980, 0, 0, 4, 56, 32, 0
from epoc:0
100000 =  1980, 0, 1, 16, 53, 0, 0
from epoc:0
1000000 =  1980, 0, 15, 8, 18, 0, 0
from epoc:0
10000000 =  1980, 4, 24, 18, 52, 0, 0
from epoc:0
62135596800000000ULL =  2107, 9, 11, 24, 0, 0, 0
Checking sceRtcGetDosTime
from epoc:0
 2107, 9, 11, 24, 0, 0, 0
ticks: 4281057280
from epoc:-1
from epoc:-1
Checking sceRtcSetWin32FileTime
  Zero time (1601 January 01): (00000000) 1601, 1, 1, 0, 0, 0, 0
  Arbitrary date/time: (00000000) 2005, 12, 1, 13, 1, 0, 1

Checking sceRtcGetWin32FileTime
  NULL filetime: -1337 (800001fe)
  Zeroed time: 0 (800001fe)
  Arbitrary date/time: 127779156600000010 (00000000)
  1601 January 01: 0 (00000000)
  1600 January 01: 0 (800001fe)
  1 January 01: 0 (800001fe)

Checking sceRtcGetTick
Normal: 00000000
Ticks : 283996800000000
Bad date: 00000000
Ticks : 315474825600000000
Min year: 800001fe
Ticks : 315474825600000000
Max year: 00000000
Ticks : 315506361600000000
Year overflow: 800001fe
Ticks : 315506361600000000
checkSetTick: empty small value: 00000000
 1, 1, 1, 0, 0, 0, 835072
checkSetTick: empty rtcMagicOffset: 00000000
 1970, 1, 1, 0, 0, 0, 0
Normal: 00000000
checkSetTick: 2012, 09, 20, 7, 12, 15, 500: 00000000
 2012, 9, 20, 7, 12, 15, 500
preset
 2010, 9, 20, 7, 12, 15, 500
checkSetTick: not empty:00000000
 2012, 9, 20, 7, 12, 15, 500
Normal:
 2012, 1, 30, 2, 12, 15, 900
RFC 2822: Mon, 30 Jan 2012 02:12:15 +0000
Normal:
 2010, 9, 20, 7, 12, 15, 500
RFC 2822: Mon, 20 Sep 2010 07:12:15 +0000
Normal:
 2012, 1, 30, 2, 12, 15, 900
RFC 3339: 2012-01-30T02:12:15.00Z
Normal:
 2010, 9, 20, 7, 12, 15, 500
RFC 3339: 2010-09-20T07:12:15.00Z

sceRtcParseDateTime:
  From RFC3339 (Z): 00000000: 2010-09-20T07:12:15.00Z -> 2010-09-20T07:12:15.00Z
  From RFC3339 (offset): 00000000: 2010-09-20T07:12:15.00+01:00 -> 2010-09-20T06:12:15.00Z
  From RFC3339 (no ms): 00000000: 2010-09-20T07:12:15Z -> 2010-09-20T07:12:15.00Z
  Bad RFC3339 (missing TZ): ffffffff: 2010-09-20T07:12:15 -> 0001-01-01T00:00:00.00Z
  Space separator: ffffffff: 2010-09-20 07:12:15Z -> 0001-01-01T00:00:00.00Z
  Human speak: ffffffff: today -> 0001-01-01T00:00:00.00Z
  Time only: ffffffff: 07:12:15 -> 0001-01-01T00:00:00.00Z
  Date only: ffffffff: 2010-09-20 -> 0001-01-01T00:00:00.00Z
  From RFC2822: 00000000: Mon, 30 Jan 2012 02:12:15 +0000 -> 2012-01-30T02:12:15.00Z
  From RFC2822 date: ffffffff: Mon, 30 Jan 2012 -> 0001-01-01T00:00:00.00Z
  From RFC2822 full weekday: 00000000: Monday, 30 Jan 2012 02:12:15 +0000 -> 2012-01-30T02:12:15.00Z
  From RFC2822 no weekday: ffffffff: 30 Jan 2012 02:12:15 +0000 -> 0001-01-01T00:00:00.00Z
  From RFC2822 + spaces: ffffffff: Mon,  30  Jan  2012  02:12:15  +0000 -> 0001-01-01T00:00:00.00Z
  Blank: ffffffff:  -> 0001-01-01T00:00:00.00Z
